proposals/001-ClusterRepresentation.md

```diff
@@ -0,0 +1,305 @@
+
+An _upstream cluster_ may reference a `filter chain`.  This provides zero or more *additional* filters that the RPCs will pass through before passing to the brokers
+of the physical cluster.  
+
+An _upstream cluster_ enumerates the upstream brokers that comprise the physical cluster.
+
+#### Upstream Brokers
+
+Each upstream broker specifies the address of the physical broker and whether TLS is to be used.  It will also specify trust settings.
+
+TLS trust material may be provided at the upstream broker level or at the individual broker level.
+
+```
+- name: my-private-cluster
+  filterChainRef: my-downstream-filter-chain
+  tls:
+    trust:
+    ...
+  brokers:
+  - name: broker-1
+    address: broker-1-private:9092
+    ...
+    tls:
+      trust:
+      ...
+  - name: broker-2
+    address: broker-1-private:9092
+    ...
+    tls:
+      trust:
+```
+
+#### Filter Chain
+
+Currently kroxylicious provide a single filter chain.  It will refactored so that we can support many filter chains, each identified by name. The `downstream cluster` and `upstream cluster` may reference a filter chain. 
+
+```yaml
+filterChains:
+- name: my-upstream-filter-chain
+  filters:
+  - type: ApiVersions
+  - type: BrokerAddress
+
+```
+
+
+### Responsibilities
+
+All specified endpoints will be bound.
+
+When a connection is made to an endpoint,  the system must resolve that connection to a downstream broker and hence a downstream cluster.
+
+To do this, it resolves the endpoint and any SNI information against the model.   This should yield exactly one broker belonging to a downstream cluster.
+It is an error otherwise and the connnection must be closed.
+
+If TLS is in use, the SSLContext can be generated from the downstream cluster definition.  This will be passed to Netty to let it complete the TLS handshake.
+
+The downstream broker and downstream cluster is used to identify the upstream broker and upstream cluster.
+
+The upstream cluster and downstream broker provide the filter chain.
+
+The handler chain connects to the upstream broker.
+
+The system must reload chnages to the model dynammically, without dropping established connections.
+
+### Broker Address Filter
+
+How will broker address filter map the RPCs that contain the upstream broker address to the downstream broker addresss that are resolveable to the client?
+
+In the case where kroxy is being placed in front of a kafka cluster spanning a three-AZ cluster, the kroxylicious instances in the AZ won't need configuration to
+*connect* to brokers in the other AZs.  However, BrokerAddressFilter will be to be capable of rewriting the the broker address for the whole cluster.
+This leads us to the conclusion that Broker Address filter need separate configuration independent of that what could be derived from the upstream/downstream cluster mapping.
+
+TODO: Maybe used a regexp based mapping would be sufficient
```

Comment:
Yes I believe a mapping will work.  The mapping will need the port numbers.  There's the possibility that downstream and upstream port numbers would be different.

I've been thinking about where this mapping lives.   If it is exists as filter configuration, then in the multitenant case, we'd need a separate filter configuration for each tenant.  Whilst that would work, I'm wondering whether the cluster configuration should have have a map of arbitrary configuration that is made available to the filters at runtime (somehow).  

We could use this to pass extra information to the filters like the MultiTenantFilter (a tenant key) and BrokerAddressFilter (a mapping).

```yaml
- name: tenant1-cluster
  upstreamClusterRef: my-private-cluster
  filterChainRef: my-upstream-filter-chain
  filterVariables:
      io.kroxylicious.tenantKey: xyz4a
      io.kroxylicious.brokerAddressMapping: 
        broker-1.azA.internal.svc:1234: broker-A.myvanitydomain.com:9080
```

I think some of the variable name would be 'well-known' (we reserve names variable names beginning with `io.kroxylicious`.

Reply:
I was assuming we would generate the mapping table from the configuration rather than listing it explicitly as we are currently talking about having exhaustive mappings. 

But yeah maybe including it in the config is just simpler and easier to deal with for now.

Reply:
> I was assuming we would generate the mapping table from the configuration rather than listing it explicitly as we are currently talking about having exhaustive mappings.

That was my initial thinking too.  I don't think that works. 

- Firstly, in the case where you have an AZ specific kroxy deployment, the mapping needs to know about the brokers on the other AZs.
- Secondly, there's no guarantee that the endpoint address/port that kroxy needs to bind is the same as the address/port that actually needs to be advertised to the client.

Reply:
I'm now passing separate configuration to BrokerAddressFilter